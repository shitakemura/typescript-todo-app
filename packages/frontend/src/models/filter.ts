export const FILTER_VALUES = ["ALL", "COMPLETED", "NOT COMPLETED"] as const;
type FilterTupel = typeof FILTER_VALUES;
export type Filter = FilterTupel[number];
