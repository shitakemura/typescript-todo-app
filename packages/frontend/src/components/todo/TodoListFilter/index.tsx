import { Button, HStack } from "@chakra-ui/react";
import { Filter, FILTER_VALUES } from "../TodoList";

type Props = {
  filter: Filter;
};

const TodoListFilter = ({ filter }: Props) => {
  return (
    <HStack spacing={4}>
      {FILTER_VALUES.map((filterValue) => {
        return (
          <Button
            key={filterValue}
            colorScheme={filter === filterValue ? "teal" : undefined}
            color={filter === filterValue ? "white" : undefined}>
            {filterValue}
          </Button>
        );
      })}
    </HStack>
  );
};

export default TodoListFilter;
