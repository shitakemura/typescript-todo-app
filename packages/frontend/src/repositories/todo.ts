import { Todo, TodoBody } from "@/shared/models/todo";
import { useAuth0 } from "@auth0/auth0-react";
import { deleteApi, getApi, postApi, putApi } from "src/api";

export const useTodoRepository = () => {
  const baseUrl = process.env.NEXT_PUBLIC_API_ENDPOINT_URL!;
  const { getAccessTokenSilently } = useAuth0();

  const listTodos = async (): Promise<Todo[]> => {
    const accessToken = await getAccessTokenSilently({
      audience: process.env.NEXT_PUBLIC_AUTH0_API_AUDIENCE,
      scope: process.env.NEXT_PUBLIC_AUTH0_OPENID_CONNECT_SCOPE,
    });
    const response = await getApi(`${baseUrl}/todos`, accessToken);
    if (!response.ok) {
      throw new Error(response.statusText);
    }
    const data = await response.json();
    return data as Todo[];
  };

  const createTodo = async (body: TodoBody): Promise<Todo> => {
    const accessToken = await getAccessTokenSilently({
      audience: process.env.NEXT_PUBLIC_AUTH0_API_AUDIENCE,
      scope: process.env.NEXT_PUBLIC_AUTH0_OPENID_CONNECT_SCOPE,
    });
    const response = await postApi(`${baseUrl}/todos`, accessToken, body);
    if (!response.ok) {
      throw new Error(response.statusText);
    }
    const data = await response.json();
    return data as Todo;
  };

  const updateTodo = async (id: string, body: Todo): Promise<Todo> => {
    const accessToken = await getAccessTokenSilently({
      audience: process.env.NEXT_PUBLIC_AUTH0_API_AUDIENCE,
      scope: process.env.NEXT_PUBLIC_AUTH0_OPENID_CONNECT_SCOPE,
    });
    const response = await putApi(`${baseUrl}/todos/${id}`, accessToken, body);
    if (!response.ok) {
      throw new Error(response.statusText);
    }
    const data = await response.json();
    return data as Todo;
  };

  const deleteTodo = async (id: string): Promise<void> => {
    const accessToken = await getAccessTokenSilently({
      audience: process.env.NEXT_PUBLIC_AUTH0_API_AUDIENCE,
      scope: process.env.NEXT_PUBLIC_AUTH0_OPENID_CONNECT_SCOPE,
    });
    const response = await deleteApi(`${baseUrl}/todos/${id}`, accessToken);
    if (!response.ok) {
      throw new Error(response.statusText);
    }
    const data = await response.json();
    return data;
  };

  return {
    listTodos,
    createTodo,
    updateTodo,
    deleteTodo,
  };
};
