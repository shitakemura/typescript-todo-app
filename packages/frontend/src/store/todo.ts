import { atom, selector } from "recoil";
import { Todo } from "@/shared/models/todo";
import { Filter } from "src/models/filter";

export const todoListState = atom({
  key: "todoListState",
  default: [] as Todo[],
});

export const todoListFilterState = atom({
  key: "todoListFilterState",
  default: "ALL" as Filter,
});

export const filteredTodoListState = selector({
  key: "filtredTodoListState",
  get: ({ get }) => {
    const filter = get(todoListFilterState);
    const list = get(todoListState);

    switch (filter) {
      case "ALL":
        return list;
      case "COMPLETED":
        return list.filter((item) => item.completed);
      case "NOT COMPLETED":
        return list.filter((item) => !item.completed);
    }
  },
});
