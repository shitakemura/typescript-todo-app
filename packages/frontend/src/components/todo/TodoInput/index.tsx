import React, { useCallback, useEffect, useState } from "react";
import { useSetRecoilState } from "recoil";
import { Button, HStack, Input, useToast } from "@chakra-ui/react";
import { TodoBody } from "@/shared/models/todo";
import { todoListState } from "src/store/todo";
import { useCreateTodo } from "src/usecases/todo/useCreateTodo";

const TodoInput = () => {
  const { isLoading, error: createError, createTodo } = useCreateTodo();
  const [inputValue, setInputValue] = useState("");
  const setTodoList = useSetRecoilState(todoListState);
  const resetInputValue = useCallback(() => setInputValue(""), []);
  const toast = useToast();

  const handleOnChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      setInputValue(event.target.value);
    },
    []
  );

  const handleAddTodo = useCallback(async () => {
    const body: TodoBody = { title: inputValue, completed: false };
    const newTodo = await createTodo(body);
    if (newTodo) {
      setTodoList((oldTodoList) => [...oldTodoList, newTodo!]);
      resetInputValue();
    }
  }, [inputValue, createTodo, setTodoList, resetInputValue]);

  useEffect(() => {
    if (createError) {
      toast({ title: createError.message, status: "error", isClosable: true });
    }
  }, [createError, toast]);

  return (
    <HStack spacing={6}>
      <Input
        borderColor='teal.500'
        borderWidth={2}
        height={12}
        value={inputValue}
        onChange={handleOnChange}
      />
      <Button
        paddingX={8}
        colorScheme='teal'
        isLoading={isLoading}
        onClick={handleAddTodo}>
        Add Todo
      </Button>
    </HStack>
  );
};

export default TodoInput;
