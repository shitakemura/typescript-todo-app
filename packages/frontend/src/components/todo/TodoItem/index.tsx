import { useCallback } from "react";
import { useRecoilState } from "recoil";
import { Checkbox, HStack, IconButton, Text } from "@chakra-ui/react";
import { DeleteIcon } from "@chakra-ui/icons";
import { Todo } from "@/shared/models/todo";
import { todoListState } from "../../../store/todo";
import { useUpdateTodo } from "../../../usecases/todo/useUpdateTodo";
import { useDeleteTodo } from "../../../usecases/todo/useDeleteTodo";

type Props = {
  todo: Todo;
};

const TodoItem = ({ todo }: Props) => {
  const { updateTodo } = useUpdateTodo();
  const { isLoading, deleteTodo } = useDeleteTodo();
  const { userId, id, title, completed } = todo;
  const [todoList, setTodoList] = useRecoilState(todoListState);

  const handleToggleTodo = useCallback(async () => {
    const body: Todo = { userId, id, title, completed: !completed };
    const updatedTodo = await updateTodo(id, body);

    const newList = todoList.map((item) => {
      if (item.id === id) {
        return updatedTodo!;
      } else {
        return item;
      }
    });
    setTodoList(newList);
  }, [userId, id, title, completed, todoList, updateTodo, setTodoList]);

  const handleDeleteTodo = useCallback(async () => {
    await deleteTodo(id);
    const newList = todoList.filter((item) => item.id !== id);
    setTodoList(newList);
  }, [id, todoList, deleteTodo, setTodoList]);

  return (
    <HStack
      borderColor='teal.300'
      borderWidth={1}
      p={8}
      w='full'
      height={16}
      justify='space-between'
      spacing={8}>
      <HStack spacing={8}>
        <Checkbox size='lg' isChecked={completed} onChange={handleToggleTodo} />
        <Text
          textDecoration={completed ? "line-through" : undefined}
          color={completed ? "gray.500" : "black"}>
          {title}
        </Text>
      </HStack>
      <IconButton
        aria-label='delete'
        color='teal.500'
        backgroundColor='white'
        p={4}
        boxSize={5}
        isLoading={isLoading}
        icon={<DeleteIcon />}
        onClick={handleDeleteTodo}
      />
    </HStack>
  );
};

export default TodoItem;
