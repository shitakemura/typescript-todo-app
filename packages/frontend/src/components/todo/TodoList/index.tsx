import { useEffect } from "react";
import { useRecoilValue, useSetRecoilState } from "recoil";
import { VStack, List, ListItem } from "@chakra-ui/react";
import { filteredTodoListState, todoListState } from "src/store/todo";
import { useListTodos } from "src/usecases/todo/useListTodos";
import TodoInput from "../TodoInput";
import TodoListFilter from "../TodoListFilter";
import TodoItem from "../TodoItem";

const TodoList = () => {
  const { data } = useListTodos();
  const setTodoList = useSetRecoilState(todoListState);
  const todoList = useRecoilValue(filteredTodoListState);

  useEffect(() => {
    if (!data) return;
    setTodoList(data);
  }, [data, setTodoList]);

  return (
    <VStack alignItems='stretch' paddingX={[8, 24, 36, 64]} spacing={12}>
      <TodoInput />
      <TodoListFilter />
      <List w='full' paddingX={8}>
        {todoList.map((todo) => (
          <ListItem key={todo.id} paddingY={2}>
            <TodoItem todo={todo} />
          </ListItem>
        ))}
      </List>
    </VStack>
  );
};

export default TodoList;
