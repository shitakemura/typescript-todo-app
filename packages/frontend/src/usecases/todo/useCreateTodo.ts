import { Todo, TodoBody } from "@/shared/models/todo";
import { useState } from "react";
import { useTodoRepository } from "repositories/todo";

export const useCreateTodo = () => {
  const todoRepository = useTodoRepository();
  const [error, setError] = useState<Error | null>(null);

  const createTodo = async (body: TodoBody): Promise<Todo | undefined> => {
    try {
      const todo = await todoRepository.createTodo(body);
      return todo;
    } catch (error: any) {
      if (error instanceof Error) {
        setError(error);
      } else {
        setError(new Error(`Failed when create todo`));
      }
    }
  };
  return { error, createTodo };
};
