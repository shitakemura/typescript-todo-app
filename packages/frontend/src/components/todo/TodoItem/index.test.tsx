import { render, screen } from "@testing-library/react";
import { Todo } from "@/shared/models/todo";
import TodoItem from "./index";
import { RecoilRoot } from "recoil";

let todo: Todo = {
  userId: "user001",
  id: "id003",
  title: "test todo title",
  completed: true,
};

const wrapper = ({ children }: { children: React.ReactNode }) => {
  return <RecoilRoot>{children}</RecoilRoot>;
};

describe("TodoItem", () => {
  test("Snapshotテスト", () => {
    const { container } = render(<TodoItem todo={todo} />, { wrapper });
    expect(container).toMatchSnapshot();
  });

  test("初期表示", () => {
    render(<TodoItem todo={todo} />, { wrapper });

    expect(screen.getByRole("checkbox")).toBeInTheDocument();
    expect(screen.getByText("test todo title")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "delete" })).toBeInTheDocument();
  });
});
