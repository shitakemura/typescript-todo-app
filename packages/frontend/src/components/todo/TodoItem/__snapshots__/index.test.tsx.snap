// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem Snapshotテスト 1`] = `
<div>
  <div
    class="chakra-stack css-12dsoc2"
  >
    <div
      class="chakra-stack css-16awuyr"
    >
      <label
        class="chakra-checkbox css-1uiwwan"
        data-checked=""
      >
        <input
          checked=""
          class="chakra-checkbox__input"
          style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
          type="checkbox"
          value=""
        />
        <span
          aria-hidden="true"
          class="chakra-checkbox__control css-dnty2r"
          data-checked=""
        >
          <div
            style="display: flex; align-items: center; justify-content: center; height: 100%; transform: none;"
          >
            <svg
              class="css-o8vbdw"
              style="fill: none; stroke-width: 2; stroke: currentColor; stroke-dasharray: 16; opacity: 1; stroke-dashoffset: 0;"
              viewBox="0 0 12 10"
            >
              <polyline
                points="1.5 6 4.5 9 10.5 1"
              />
            </svg>
          </div>
        </span>
      </label>
      <p
        class="chakra-text css-1sb2pv9"
      >
        test todo title
      </p>
    </div>
    <button
      aria-label="delete"
      class="chakra-button css-qc4ta9"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M19.452 7.5H4.547a.5.5 0 00-.5.545l1.287 14.136A2 2 0 007.326 24h9.347a2 2 0 001.992-1.819L19.95 8.045a.5.5 0 00-.129-.382.5.5 0 00-.369-.163zm-9.2 13a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zm5 0a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zM22 4h-4.75a.25.25 0 01-.25-.25V2.5A2.5 2.5 0 0014.5 0h-5A2.5 2.5 0 007 2.5v1.25a.25.25 0 01-.25.25H2a1 1 0 000 2h20a1 1 0 000-2zM9 3.75V2.5a.5.5 0 01.5-.5h5a.5.5 0 01.5.5v1.25a.25.25 0 01-.25.25h-5.5A.25.25 0 019 3.75z"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;
